### ISSUE TITLE
DSC Esent Database Devices.edb growing and reporting slow

### ISSUE THREAD

**Issue by skest3qc** [PowerShell/xPSDesiredStateConfiguration#296]
Hi,

I am having the following issue. 
Pullserver is Server 2016 14393.477
I have registered around 80 nodes. I see that after a month of using the server the Devices.edb database is around 1.8GB and also that it takes ages to get a node report for a configuration id.

Code for getting status report (from MSDN):
$serviceURL ="PullServerURL"
$requestUri = "$serviceURL/Node(ConfigurationId='$ConfigurationId')/StatusReports"

$request = Invoke-WebRequest -Uri $requestUri  -ContentType "application/json;odata=minimalmetadata;streaming=true;charset=utf-8" `
           -UseBasicParsing -Headers @{Accept = "application/json";ProtocolVersion = "2.0"} `
           -ErrorAction SilentlyContinue -ErrorVariable ev -Verbose: $false
$object = ConvertFrom-Json $request.content
return $object.value

In the beginning i had no problems to get a report, but now after a month it takes ower 2 hours to get a report, also w3p.exe (PSWS) uses about 3GB and has high CPU usage.

I have copied Devices.edb and opened it ESEDDatabaseView and see that there are 65928 rows in StatusReport table.

Is DSC cleaning up the table automatically after some time, or is there a way to delete the rows in the StatusReport table.

Also is this a known issue? Is there a way to get results from StatusReports faster?

Thanks

**Comment by usrme** [comment]
As much as I've used DSC, it does no clean up to the *Devices.edb* file and just keeps growing with every check the LCM does against the pull server. I found no way to delete any rows from the various tables inside *Devices.edb* either.

@jordan-rogers-ge mentioned this as well in issue #201, so you could say it is somewhat known, but more attention is definitely needed. To my knowledge, there is no way to get status reports faster since for every bit of information you are querying that database file and it seems very poorly optimized for its potential size.

**Comment by J0F3** [comment]
I have exactly the same issue. Except that I have registered about 150 nodes and the database with its logfiles is now around 46GB big!
This an unacceptable situation and makes the DSC Pullserver or more specific the reporting functionality of the DSC Pullserver absolutely useless for a production environment in real world. 👎 :(

Is there really no solution for cleaning up the tables? 
I am also seeing a lot of edb00xy.log files in the same directory. Is that normal? When will the logfiles written back to the database? 

Any help is much appreciated as our Pullserver is nearly unusable in this state!

Thanks!

**Comment by J0F3** [comment]
Hi, 

Thanks for your input. I have also implemented the workaround with the scheduled task for now.
After investigating further in this issue, I found different reasons for the unexpected behavior of the Pullserver:

1. The "Microsoft.Powershell.DesiredStateConfiguration.Service.dll" is mounting the EDB File without the CircularLog parameter. Which I think is a bug. That's one reason for the huge count of log files.

2. When the Application Pool recycles, the DSCPullserver does a cleanup on the report data which are older than 30 days. As this takes some time it can happen that there are two w3wp.exe process are trying to access the EDB file simulations. When that happens the DSCPullserver responds to every request with an error message that no registered client can be found and the registration of new clients will also fail. To work around this issue, I enabled the "Disable Overlapped Recycle" setting on IIS the app pool.

3. When a DSC client reports to the server it makes a huge amount of web request. (about 10-15 per second). For me that not looks correct. But I don’t know what is the reason for that. However, I think this is maybe another reason for the huge size and count of log files of the EDB database.
4. Also during the reporting of a DSC client the w3wp.exe Process of the DSCPullserver has a lot of exceptions. “E0434F4D.System.NotImplementedException (“Complex type MSFT.PropertyBag is not supported.”)” and “E0434F4D.System.Collections.Generic.KeyNotFoundException (“The given key was not present in the directory”)”.
It looks like that therefore the status reports will not get correctly saved to the database. (this still happens with WMF 5.1). So, when I request the status reports for a client  the result is always empty.

In short, the DSCPullserver in WMF 5.1 has unfortunately still some serious bugs… :-(

**Comment by skest3qc** [comment]
I have removed PSPullServer and installed the last CU Update on Windows Server 2016.
I have redployed the PullServer and now have a different Microsoft.Powershell.DesiredStateConfiguration.Service.dll V:10.0.14393.479 (in C:\inetpub\wwwroot\PSDSCPullServer\bin)

After some testing it seems that the log files are being cleared correctly from DSC Service.

**Comment by J0F3** [comment]
That's exactly what I have discovered also today. So MSFT has apparently fixed it senility. 
I can also confirm that circular logging is enabled on the EDB if the latest version of the dll is used (see last line)

```
PS D:\DSCPullService\Database> esentutl.exe /mk "D:\DSCPullService\Database\edb.chk"

Extensible Storage Engine Utilities for Microsoft(R) Windows(R)
Version 10.0
Copyright (C) Microsoft Corporation. All Rights Reserved.

Initiating FILE DUMP mode...
      Checkpoint file: D:\DSCPullService\Database\edb.chk

      LastFullBackupCheckpoint: (0x0,0,0)
      Checkpoint: (0xE,1,0)
      DbConsistency: (0x10,10,0)
      FullBackup: (0x0,0,0)
      FullBackup time: 00/00/1900 00:00:00.000
      IncBackup: (0x0,0,0)
      IncBackup time: 00/00/1900 00:00:00.000
      Signature: Create time:01/09/2017 14:12:13.965 Rand:2646131952 Computer:
      Env (CircLog,Session,Opentbl,VerPage,Cursors,LogBufs,LogFile,Buffers)
==>     (    on,     16,    300,   1024,   1024,    126,     32,2000000000)
```

I have simply stop the Website and the App Pool and then replaced the Microsoft.Powershell.DesiredStateConfiguration.Service.dll and Microsoft.Powershell.DesiredStateConfiguration.Service.Resources.dll with the ones from C:\Windows\System32\WindowsPowerShell\v1.0\Modules\PSDesiredStateConfiguration\PullServer (Version 10.0.14393.479)

Afterward the circular logging is enabled and the size of database with its logfiles stays at the range of about 300MB - 400 MB.

**Comment by kwirkykat** [comment]
@skest3qc @J0F3  So this is fixed? Or still an issue?

**Comment by J0F3** [comment]
The part with the big EDB is now partially fixed with the latest windows updates. Meaning it is only fixed for new setups on Server 2016. Existing setups get the update not automatically because windows updates does only replaces the dll in the windows directory but not the one which the DSC Pullserver is actually using. Moreover the update of the dll is nowhere documented. So I found the solution more by accident!

Still open is the question how the DSC Pullserver can be made highly available. -> https://github.com/PowerShell/PowerShell/issues/2978 and https://github.com/PowerShell/PowerShell-Docs/issues/908

**Comment by kwirkykat** [comment]
@J0F3 Ok. I will leave this issue open then.

**Comment by BerheAbrha** [comment]
This issue is fixed on latest WMF 5.1 or 2016  (as @skest3qc  explained it above).  Now you can turn ON or OFF the circular logging behavior of the ESNET database using web.config parameter of the database. ON is the default value for this parameter.

**Comment by J0F3** [comment]
@BerheAbrha Yes, thats correct. But it has been fixed silently and only for new setups. For existing setups the issue still existis until the DSC Pullserver dll is manually replaced. IMHO that is no good practice. At least, there should exist a changelog somewhere. Otherwise, every month, it's pure luck for us what will be fixed or not after installing the latest updates!

**Comment by BerheAbrha** [comment]
@J0F3  I will follow up on the documentation process. **You are right**, the WMF 5.1 release note https://msdn.microsoft.com/en-us/powershell/wmf/5.1/dsc-improvements 
does not mention the pull server bug fix. Regarding old pull server setups, only the new code has the feature to create circular ESNET database instances. So you will need to update your pullserver binaries to get this feature.

**Comment by Zuldan** [comment]
Now that the log size issue is resolved I'd like to bring the conversation back to the speed of retrieving reports.

> In the beginning i had no problems to get a report, but now after a month it takes ower 2 hours to get a report, also w3p.exe (PSWS) uses about 3GB and has high CPU usage.

I'm having th exact same issue. I have about 200 nodes registered with the DSC Pull server (freshly build with 5.1 RTM) that perform a consistency check every 60 minutes. After 2 days it's pretty much impossible to retrieve a report for a node. The memory and CPU goes through the roof. I have a scheduled task that deleted the event DB every 24 hours to make the DB usable again. Is everyone else doing this? Has some one found a work around? Would be great if someone created a PS wrapper about the event database and queried it directly.

Jeff Hicks appears to be using a odata filter here https://www.petri.com/how-to-query-the-dsc-compliance-server. See near end of blog post. Has anyone tried to get this working?

**Comment by Zuldan** [comment]
Another idea is to use a SQL Server on the backend and then query the SQL Server directly for reports. Here are 2 slight different solutions to using a SQL Server with DSC Pull Server.

Solution 1: https://leandrowpblog.wordpress.com/2016/10/26/using-sql-server-db-for-dsc/
Solution 2: https://blogs.msdn.microsoft.com/wushuai/2016/11/27/try-to-use-sql-server-as-dsc-pull-server-backend/

**Comment by gigi81** [comment]
Hi @Zuldan ,
I investigated the same "status report slowness" issue. On our server it takes roughly 30 seconds to get the status report for one node. Considering we have 88, it's about 45 minutes in total to get all statuses, while the server's CPU spikes to 50% to 100% and IIS process uses from 1GB to 2GB.
I imported the DSC odata api into a C# project as a service reference and tried to play with the different Linq queries you can create.
Long story short, I confirm there is no way to get decent performance out of it.
These are the things I tried:
- using directly the /StatusReport endpoint doesn't work (GET is not supported)
- using the /Nodes endpoint times out (server error)
- tried to delete the StatusReports through webapi but getting an error that the method is not implemented
- using the /Node(ConfigurationId='guid')/StatusReports() is the only one working but it doesn't support the $top so the slowness performance

I also tried to open the ESENT database directly but I can't because it's in use by the DSC service.

The only option I can think of for the moment is to completely wipe out the esent database every night.

**Comment by skest3qc** [comment]
I am currently using this module which is working fine:
[DSCPullServerAdmin](https://github.com/bgelens/DSCPullServerAdmin)

I am also deleting rows from StatusReports table with this module

**Comment by gigi81** [comment]
Hi @skest3qc 
Thanks for the suggestion, it is actually a good one. Anyway, as I was saying before, you cannot open the esent db while it is in use by the pull server. This is the error I get:

```powershell
Mount-DSCPullServerAdminDatabase -ESEPath "C:\Program Files\WindowsPowerShell\DscService\Devices.edb"
Mount-DSCPullServerAdminDatabase : Cannot access file, the file is locked or in use
At line:1 char:1
+ Mount-DSCPullServerAdminDatabase -ESEPath "C:\Program Files\WindowsPo ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OpenError: (C:\Program File...ice\Devices.edb:String) [Mount-DSCPullServerAdminDatabase], EsentFileAccessDeniedException
    + FullyQualifiedErrorId : Cannot access file, the file is locked or in use,DSCPullServerAdmin.src.CmdLets.MountDSCPullServerDatabase
```

**Comment by skest3qc** [comment]
Yes just stop the iis app pool and after getting the report start it again...

**Comment by J0F3** [comment]
Exactly the EDB is only editable after the app pool is stopped.

All in all it seems that the switch to the ESENT DB was not really a good idea. It makes also problems when you want to make the pull server higly availabe. Or if you want to assign a different configuration to a existing pull client.

A simple SQL DB would probably made everything easier....

**Comment by dotps1** [comment]
After setting up a pull server, my ESENT DB is not growing out of control, I am however getting 100's of 1000's of edbXXXX.log files every day.  I have found blogs about switching to a JET OLEDB provider and using the .mdb, but on Server 2016 Core I have not been able to get that to work.

**Comment by skest3qc** [comment]
Install all OS patches and do this:
 Simply stop the Website and the App Pool and then replace the Microsoft.Powershell.DesiredStateConfiguration.Service.dll and Microsoft.Powershell.DesiredStateConfiguration.Service.Resources.dll with the ones from C:\Windows\System32\WindowsPowerShell\v1.0\Modules\PSDesiredStateConfiguration\PullServer (Version 10.0.14393.479)

**Comment by dotps1** [comment]
says there is no patches to install, and that details tab of `C:\Windows\System32\WindowsPowerShell\v1.0\Modules\PSDesiredStateConfiguration\PullServer\Microsoft.Powershell.DesiredStateConfiguration.Service.Resources.dll` is `10.0.14393.0`.  Any idea what KB replaces that .dll?

Thanks!

**Comment by dotps1** [comment]
Ok, so I'm just going to put this in here incase anyone else runs into this.

mind you, I'm running Server 2016 Core x64:
* RDP to the machine.
* `get-item C:\Windows\System32\WindowsPowerShell\v1.0\Modules\PSDesiredStateConfiguration\PullServer\Microsoft.Powershell.DesiredStateConfiguration.Service.dll | select -ExpandProperty VersionInfo | Select -ExpandProperty ProductVersion`
* 10.0.14393.0
* run `sconfig`.
* check for updates.
* use option 'A' to check for "all updates".
* no results.
* Install PSWindowsUpdate module from the PSGallery `Install-Module PSWindowsUpdate`.
* Check for updates with PSWindowsUpdate `Get-WindowsUpdate`.
* no results.
* Started looking at parameters of `Get-WindowsUpdate`, noticed the -MicrosoftUpdate.
* run command `Get-WindowsUpdate -MicrosoftUpdate`.
* get an error: `WARNING: Can't find registered service Microsoft Update. Use Get-WUServiceManager to get registered service.`
* googles
* found this blog: https://www.windowsbbs.com/threads/powershell-for-automatize-windows-update-and-program-installation.110303/
* register the Microsoft Update Service Manager: `Add-WUServiceManager -ServiceID 7971f918-a847-4430-9279-4a52d1efe18d -Confirm:$false`
* Check for updates: `Get-WUInstall -MicrosoftUpdate`
* Finds four updates, including `KB4015217` which is a cumulative update for Server 2016.
* Install update `Install-WindowsUpdate -KBArticleID KB4015217 -MicrosoftUpdate`
* Reboot to finish.
* `get-item C:\Windows\System32\WindowsPowerShell\v1.0\Modules\PSDesiredStateConfiguration\PullServer\Microsoft.Powershell.DesiredStateConfiguration.Service.dll | select -ExpandProperty VersionInfo | Select ProductVersion`
* 10.0.14393.479

FML!

**Comment by johlju** [comment]
Since there will no more features to Windows Server Pull Server, referencing [Windows Pull Server Planning Update – April 2018](https://blogs.msdn.microsoft.com/powershell/2018/04/19/windows-pull-server-planning-update-april-2018/) I'm closing this as external since this cannot be resolved in this resource module.
